router/pkg/authentication/jwks_token_decoder.go

```diff
@@ -46,16 +47,26 @@ type JWKSConfig struct {
 	Secret    string
 	Algorithm string
 	KeyId     string
-}
 
-func NewJwksTokenDecoder(ctx context.Context, logger *zap.Logger, configs []JWKSConfig) (TokenDecoder, error) {
+	Audiences []string
+}
 
-	remoteJWKSets := make(map[string]jwkset.Storage)
+type AudKey struct {
+	kid string
+	url string
+}
 
-	given := jwkset.NewMemoryStorage()
+func NewJwksTokenDecoder(ctx context.Context, logger *zap.Logger, configs []JWKSConfig) (TokenDecoder, error) {
+	audiencesMap := make(map[AudKey][]string, len(configs))
```

Comment:
We could optimize this a bit as we are iterating on every request.
Imagine something like (high level)

```go
type set[T comparable] map[T]struct{}{}
type audienceStore map[AudKey]set[string]
```

Then you can use the `set` in the `hasAudience` function directly

Reply:
Done, didn't make it a generic type though and made it a more specific type of `audienceSet`